File: ios/Anemone SDK/model/config/PropertyOption.swift
//
//  PropertyOption.swift
//  Anemone SDK
//
//  Created by Manuel Gonzalez Villegas on 9/12/16.
//  Copyright © 2016 Syltek Solutions S.L. All rights reserved.
//

import Foundation

public typealias PropertyOptionId = Id

public struct PropertyOption {
    public static let singleSizePropertyOption = PropertyOptionId("single")
    public static let doubleSizePropertyOption = PropertyOptionId("double")
    public static let miniSizePropertyOption = PropertyOptionId("mini")

    public let id: PropertyOptionId
    public let name: String
    public let filterable: Bool

    public init(id: PropertyOptionId, name: String, filterable: Bool = true) {
        self.id = id
        self.name = name
        self.filterable = filterable
    }
}

extension PropertyOption: JSONMappable {
    public init(json: JSONObject) throws {
        self.id = try PropertyOptionId(json.getAny("option_id"))
        self.name = try json.getString("name")
        self.filterable = (try? json.getBoolean("filterable")) ?? true
    }
}

extension PropertyOption: Equatable { }
public func == (lhs: PropertyOption, rhs: PropertyOption) -> Bool {
    lhs.id == rhs.id
}

File: android/app/src/main/java/com/playtomic/test/anemone_sdk/model/config/PropertyOption.kt
package com.anemonesdk.model.config

import com.anemonesdk.general.json.JSONMappable
import com.anemonesdk.general.json.JSONObject
import com.anemonesdk.model.generic.Id
import org.json.JSONException

/**
 * Created by agarcia on 23/12/2016.
 */

typealias PropertyOptionId = Id

class PropertyOption : JSONMappable {
    companion object {
        val singleSizePropertyOption = PropertyOptionId("single")
        val doubleSizePropertyOption = PropertyOptionId("double")
        val miniSizePropertyOption = PropertyOptionId("mini")
    }

    val id: PropertyOptionId
    val name: String
    val filterable: Boolean

    constructor(id: PropertyOptionId, name: String, filterable: Boolean = true) {
        this.id = id
        this.name = name
        this.filterable = filterable
    }

    @Throws(JSONException::class)
    constructor(json: JSONObject) : super(json) {
        id = PropertyOptionId(json.getAny("option_id"))
        name = json.getString("name")
        filterable = json.optBoolean("filterable") ?: true
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        return (other as? PropertyOption)?.id == id
    }

    override fun hashCode(): Int = id.hashCode()
}
